### PR TITLE
known_failure decorator tracks annotations as a list of dicts.

### PR DESCRIPTION
**Pull request by smccarthy788**
Update `collect_known_failure.py` to reflect this.

@mambocab, please review.

Tests with two annotations will be represented by `collect_known_failure.py` as:

```bash
{"failure_annotations": [{"failure_source": "test", "notes": "", "known_flaky": true, "jira_url": 
"https://issues.apache.org/jira/browse/CASSANDRA-12444"}, {"failure_source": "cassandra", "notes": "", "known_flaky": true, "jira_url": "https://issues.apache.org/jira/browse/CASSANDRA-12457"}], "name": "TestUpgrade_current_3_x_To_indev_3_x.rolling_upgrade_test", "module": "upgrade_tests.upgrade_through_versions_test"} ... ok
```

Something we should think about is any place where we access the `known_failure` attribute thinking there will be the `failure_source` value. I'm not aware of anything like this in dtest, but maybe it is used in Jenkins?

### CODE REVIEW

File: tools/decorators.py [hunk 1]
```diff
@@ -155,11 +155,20 @@ def wrapper(f):
...
+        except AttributeError:
+            existing_failure_annotations = []
+
+        new_annotation = [{'failure_source': failure_source, 'jira_url': jira_url, 'notes': notes, 'known_flaky': flaky}]
+
+        failure_annotations = existing_failure_annotations + new_annotation
+
+        tagged_func = attr(known_failure=failure_annotations)(f)
+
         if flaky:
```
Comment (mambocab) [inline tools/decorators.py]: I think we might want to remove this as well. I'll flesh my thoughts out further in a top-level comment

**Comment by mambocab** [comment]
> Tests with two annotations will be represented by collect_known_failure.py as:

I'm surprised this doesn't present as a json list.

**Comment by mambocab** [comment]
> Something we should think about is any place where we access the known_failure attribute thinking there will be the failure_source value. I'm not aware of anything like this in dtest, but maybe it is used in Jenkins?

I'm not sure I follow -- afiak, Jenkins doesn't consume this information, except maybe with the `nosetests -a` option. That's something to look into.

Speaking of `nosetests -a`, this change makes the documentation for `known_failures` incorrect, and may affect other documentation. This should be addressed.

Finally, @ptnapoleon/@knifewine, do we want to use this as an opportunity to fix the failure annotation API and names? There we could maybe improve the annotation interface for people using `nosetests -a`, though I can't think of a particularly useful application here. Is "run all the flaky tests" an interesting command? We could also remove `failure_source` and change `known_flaky` just to `flaky`.

Maybe these changes are out of scope or should be done later. I just know we've talked about some of this and it might be good to get it all out of the way in one go.

**Comment by mambocab** [comment]
> I'll dig around for other cases

Someone with ssh access to CassCI should be able to grep for it, I think.

> there's a documented case where we want to run `nosetests -a known_failure=test`

Yeah, that'd be broken here. One solution, which I don't like much, is to require clients to call something like

```sh
nosetests -A "'test' in [d['failure_source'] for d in known_failure]"
```

We could track failure sources both in the list of dicts and in a `failure_sources` attribute to make this a bit easier:

```sh
nosetests -A "'test' in failure_sources"
```

But, I don't know that we can get around using `-A` without doing something much, much grosser. (To be clear, we probably should not do something much, much grosser.) I don't know if I like doubling up that data to make it easier to query via `-a`.

Just out of curiosity, what's the use case where we run with `-a known_failure=test`? I remember talking with folks about removing the failure type annotation entirely, so I'm surprised someone's using it.

> there's an edge case where one annotation is flaky and the other isn't. Is the test flaky?

I guess it's not? If one of the errors makes it fail all the time, then it's a hard failure. So, for a given test, `flaky` should be `all(d['flaky'] for d in known_failure)`.

**Comment by smccarthy788** [comment]
>  what's the use case where we run with `-a known_failure=test`
It's just mentioned in the `known_failure` docs.

**Comment by mambocab** [comment]
> It's just mentioned in the known_failure docs.

Ah, cool -- I think that's just a usage example. Unless you know of somewhere it's

**Comment by mambocab** [comment]
Ping -- have you found any actual uses of `-a known_failure=...` or similar?

**Comment by smccarthy788** [comment]
I have found no uses of `-a known_failures=...` etc.

**Comment by mambocab** [comment]
Niiice. I think we can break the whole `-a` API then. @ptnapoleon ?

I think that also means removing the `if flaky` block that adds the extra attribute, yeah?

**Comment by ptnapoleon** [comment]
Go ahead and break the API

**Comment by smccarthy788** [comment]
@mambocab, what do you think of these changes?

**Comment by mambocab** [comment]
Code looks good! Ran it locally and I saw the line

```
{"failure_annotations": [{"failure_source": "test", "notes": "windows", "flaky": false, "jira_url": "https://issues.apache.org/jira/browse/CASSANDRA-11236"}, {"failure_source": "cassandra", "notes": "", "flaky": true,               "jira_url": "https://issues.apache.org/jira/browse/CASSANDRA-12337"}], "name": "TestScrubIndexes.test_standalone_scrub", "module": "scrub_test"} ... ok 
```

Which is what I had hoped for. Merge when you're ready.
